storage/allocator_test.go

```diff
@@ -674,6 +674,476 @@ func TestAllocatorRemoveTarget(t *testing.T) {
+					},
+					{
+						Attrs: []string{"us-east"},
+					},
+					{
+						Attrs: []string{"us-east"},
+					},
+				},
+				RangeMinBytes: 0,
+				RangeMaxBytes: 64000,
+			},
+			desc: proto.RangeDescriptor{
+				Replicas: []proto.Replica{
+					{
+						StoreID:   1,
+						NodeID:    1,
+						ReplicaID: 1,
+					},
+					{
+						StoreID:   2,
+						NodeID:    2,
+						ReplicaID: 2,
+					},
+					{
+						StoreID:   3,
+						NodeID:    3,
+						ReplicaID: 3,
+					},
+				},
+			},
+			expectAction: false,
+		},
+	}
+
+	lastPriority := float64(999999999)
+	for i, tcase := range testCases {
+		canRepair, needsRepair, priority := a.requiresRepair(tcase.zone, &tcase.desc)
+		if !canRepair {
+			t.Errorf("Test case %d indicated the range was unavailable", i)
+			continue
+		}
+		if tcase.expectAction && !needsRepair {
+			t.Errorf("Test case %d should need repair, allocator returned false.", i)
+			continue
+		}
+		if !tcase.expectAction && needsRepair {
+			t.Errorf("Test case %d should not need repair, allocator returned true.", i)
+			continue
+		}
+		if priority >= lastPriority {
+			t.Errorf("Test cases should have descending priority. Case %d had priority %f, previous case had priority %f", i, priority, lastPriority)
+		}
+		lastPriority = priority
+	}
+}
+
+func TestAllocatorComputeRepair(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	stopper, g, _, a := createTestAllocator()
+	defer stopper.Stop()
+
+	// Setup the stores so that store 3 is the worst candidate.
+	stores := []*proto.StoreDescriptor{
+		{
+			StoreID:  1,
+			Node:     proto.NodeDescriptor{NodeID: 1},
+			Capacity: proto.StoreCapacity{Capacity: 100, Available: 80, RangeCount: 10},
+			Attrs: proto.Attributes{
+				Attrs: []string{"us-east"},
+			},
+		},
+		{
+			StoreID:  2,
+			Node:     proto.NodeDescriptor{NodeID: 2},
+			Capacity: proto.StoreCapacity{Capacity: 100, Available: 60, RangeCount: 10},
+			Attrs: proto.Attributes{
+				Attrs: []string{"us-east"},
+			},
+		},
+		{
+			StoreID:  3,
+			Node:     proto.NodeDescriptor{NodeID: 3},
+			Capacity: proto.StoreCapacity{Capacity: 100, Available: 60, RangeCount: 10},
+			Attrs: proto.Attributes{
+				Attrs: []string{"us-east"},
+			},
+		},
+		{
+			StoreID:  4,
+			Node:     proto.NodeDescriptor{NodeID: 3},
+			Capacity: proto.StoreCapacity{Capacity: 100, Available: 20, RangeCount: 100},
+			Attrs: proto.Attributes{
+				Attrs: []string{"us-east"},
+			},
+		},
+	}
+
+	sg := newStoreGossiper(g)
+	sg.gossipStores(stores, t)
+
+	testzone := config.ZoneConfig{
+		ReplicaAttrs: []proto.Attributes{
+			{
+				Attrs: []string{"us-east"},
+			},
+			{
+				Attrs: []string{"us-east"},
+			},
+			{
+				Attrs: []string{"us-east"},
+			},
+		},
+		RangeMinBytes: 0,
+		RangeMaxBytes: 64000,
+	}
+
+	testCases := []struct {
+		zone            config.ZoneConfig
+		desc            proto.RangeDescriptor
+		expectErr       bool
+		expectedAction  allocatorAction
+		expectedReplica proto.Replica
+	}{
+		// Add a replica.
```

Comment:
If requiresRepair and computeRepair remain separate functions, we should at least share the same testCase data to verify that they return results that are consistent with each other.